core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1791,10 +1793,10 @@ class Log(@volatile var dir: File,
    * @param isRecoveredSwapFile true if the new segment was created from a swap file during recovery after a crash
    */
   private[log] def replaceSegments(newSegments: Seq[LogSegment], oldSegments: Seq[LogSegment], isRecoveredSwapFile: Boolean = false) {
-    val sortedNewSegments = newSegments.sortBy(_.baseOffset)
-    val sortedOldSegments = oldSegments.sortBy(_.baseOffset)
-
     lock synchronized {
+      val sortedNewSegments = newSegments.sortBy(_.baseOffset)
+      val sortedOldSegments = oldSegments.filter(seg => segments.containsKey(seg.baseOffset)).sortBy(_.baseOffset)
```

Comment:
Might be good to add a comment on why we need `containsKey`.

Reply:
Sure. I added the comment `Some old segments may have been removed from index and scheduled for async deletion after the caller reads segments but before this method is executed. We want to filter out those segments to avoid calling asyncDeleteSegment() multiple times for the same segment.`